javascript/ql/src/semmle/javascript/dataflow/Configuration.qll

```diff
@@ -699,14 +676,165 @@ private predicate isSink(DataFlow::Node nd, DataFlow::Configuration cfg, FlowLab
 }
 
 /**
- * Holds if `nd` may be reachable from a source under `cfg`.
+ * Holds if there is an exploratory flow step `pred -> succ` in `cfg`.
  *
- * No call/return matching is done, so this is a relatively coarse over-approximation.
+ * These disregard flow labels and store/load matching, but do not step
+ * in and out of functions.
+ */
+predicate basicExploratoryFlowStep(
+  DataFlow::Node pred, DataFlow::Node succ, DataFlow::Configuration cfg
+) {
+  isRelevantForward(pred, cfg, _) and
+  (
+    localFlowStep(pred, succ, cfg, _, _)
+    or
+    isAdditionalStoreStep(pred, succ, _, cfg)
+    or
+    isAdditionalLoadStep(pred, succ, _, cfg)
+    or
+    isAdditionalLoadStoreStep(pred, succ, _, cfg)
+    or
+    propertyFlowStep(pred, succ)
+    or
+    globalFlowStep(pred, succ)
+    or
+    basicStoreStep(pred, succ, _)
+    or
+    basicLoadStep(pred, succ, _)
+    or
+    callback(pred, succ)
+    or
+    succ = pred.(DataFlow::FunctionNode).getAParameter()
+  )
+}
+
+/**
+ * Gets the container depth of the given function or function input (parameter or receiver),
+ * or -1 if `nd` is the pseudo-node indicating "no parameter".
+ *
+ * The depth is used by exploratory flow to determine it can step out of a function to a
+ * given call site.
+ */
+cached
+private int getContainerDepthFromInputNode(DataFlow::Node nd) {
+  exists(Function f |
+    nd = DataFlow::parameterNode(f.getAParameter())
+    or
+    nd = DataFlow::thisNode(f)
+    or
+    nd = DataFlow::valueNode(f)
+  |
+    result = f.getContainerDepth()
+  )
+  or
+  nd = DataFlow::globalAccessPathRootPseudoNode() and
+  result = -1
+}
+
+/**
+ * Holds if `nd` is an assignment to a captured variable declared
+ * in the function `scope`.
+ */
+cached
+private predicate capturedVariableScope(DataFlow::Node nd, DataFlow::Node scope) {
```

Comment:
The point is to speed up the join by caching exactly the relation we need. Otherwise we'd need to do `DataFlow::Node.getContainer()` at every iteration of exploratory flow.

Reply:
Perhaps rename the parameter to `scopeContainer` then?